srv/apns/http_api/processor.go

```diff
@@ -0,0 +1,117 @@
+package http_api
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/uniqush/uniqush-push/push"
+	"github.com/uniqush/uniqush-push/srv/apns/common"
+)
+
+// HTTPPushRequestProcessor sends push notification requests to APNS using HTTP API
+type HTTPPushRequestProcessor struct {
+	client *http.Client
+}
+
+// NewRequestProcessor returns a new HTTPPushProcessor using net/http DefaultClient connection pool
+func NewRequestProcessor() common.PushRequestProcessor {
+	return &HTTPPushRequestProcessor{
+		client: http.DefaultClient,
+	}
+}
+
+func (processor *HTTPPushRequestProcessor) AddRequest(request *common.PushRequest) {
+	go processor.sendRequests(request)
+}
+
+func (processor *HTTPPushRequestProcessor) GetMaxPayloadSize() int {
+	return 4096
+}
+
+func (processor *HTTPPushRequestProcessor) Finalize() {
+	switch transport := processor.client.Transport.(type) {
+	case *http.Transport:
+		transport.CloseIdleConnections()
+	default:
+	}
+}
+
+func (processor *HTTPPushRequestProcessor) SetErrorReportChan(errChan chan<- push.PushError) {}
+
+func (processor *HTTPPushRequestProcessor) sendRequests(request *common.PushRequest) {
+	defer close(request.ErrChan)
+
+	jwtManager, err := common.NewJWTManager(request.PSP.FixedData["p8"], request.PSP.FixedData["keyid"], request.PSP.FixedData["teamid"])
+	if err != nil {
+		request.ErrChan <- push.NewError(err.Error())
+		return
+	}
+	jwt, err := jwtManager.GenerateToken()
+	if err != nil {
+		request.ErrChan <- push.NewError(err.Error())
+		return
+	}
+
+	wg := new(sync.WaitGroup)
+	wg.Add(len(request.Devtokens))
+
+	header := http.Header{
+		"authorization":   []string{"bearer " + jwt},
```

Comment:
Authorization header is used since at this point, the user will have already chosen to use p8 file instead of certificate. p8 refers to a path to PKCS8 key file which is different from certificate file.

Yes, I missed the part about generating new tokens, will refactor it to hourly.